src/Pirouette/Symbolic/Eval.hs

```diff
@@ -311,12 +323,30 @@ symEvalOneStep t@(R.App hd args) = case hd of
     mayBranches <- lift $ branchesBuiltinTerm @lang builtin translator args
     case mayBranches of
       -- if successful, open all the branches
-      Just branches -> asum $
-        flip map branches $ \(SMT.Branch additionalInfo newTerm) -> do
-          lift $ learn additionalInfo
-          consumeFuel
-          signalEvaluation
-          pure newTerm
+      Just branches -> do
+        -- do
+        -- consumeFuel
+        -- signalEvaluation
+        -- lift $ learnBranches branches
+        --
+        asum $
+          flip map branches $ \(SMT.Branch additionalInfo newTerm) -> do
+            lift $ learn additionalInfo
+            consumeFuel
+            signalEvaluation
+            pure newTerm
+      --
+      -- let problemsWithTerms = asum $
+      --       flip map branches $ \(SMT.Branch additionalInfo newTerm) -> do
+      --         problem <- pruneOrGetPathProblem additionalInfo
+      --         pure $ (newTerm, problem)
+      -- (newTerm, result) <- lift $ solveAll problemsWithTerms
```

Comment:
Another way to do the same thing is described is this comment. Now we first gather the problems for all the possible pairs (current state, branch constraint), and solve all these problems asynchronously in one go.